src/SFML/Window/Win32/JoystickImpl.cpp

```diff
@@ -101,6 +106,212 @@ JoystickBlacklist joystickBlacklist;
+                               reinterpret_cast<LPVOID*>(&data.pIWbemLocator));
+    if (FAILED(hr) || data.pIWbemLocator == nullptr)
+    {
+
+        return false;
+    }
+
+    data.bstrNamespace = SysAllocString(L"\\\\.\\root\\cimv2");
+    if (data.bstrNamespace == nullptr)
+    {
+        return false;
+    }
+    data.bstrClassName = SysAllocString(L"Win32_PNPEntity");
+    if (data.bstrClassName == nullptr)
+    {
+        return false;
+    }
+    data.bstrDeviceID = SysAllocString(L"DeviceID");
+    if (data.bstrDeviceID == nullptr)
+    {
+        return false;
+    }
+
+    // Connect to WMI
+    hr = data.pIWbemLocator
+             ->ConnectServer(data.bstrNamespace, nullptr, nullptr, nullptr, 0L, nullptr, nullptr, &data.pIWbemServices);
+    if (FAILED(hr) || data.pIWbemServices == nullptr)
+    {
+        return false;
+    }
+
+    // Switch security level to IMPERSONATE.
+    hr = CoSetProxyBlanket(data.pIWbemServices,
+                           RPC_C_AUTHN_WINNT,
+                           RPC_C_AUTHZ_NONE,
+                           nullptr,
+                           RPC_C_AUTHN_LEVEL_CALL,
+                           RPC_C_IMP_LEVEL_IMPERSONATE,
+                           nullptr,
+                           EOAC_NONE);
+    if (FAILED(hr))
+    {
+        return false;
+    }
+
+    hr = data.pIWbemServices->CreateInstanceEnum(data.bstrClassName, 0, nullptr, &data.pEnumDevices);
+    if (FAILED(hr) || data.pEnumDevices == nullptr)
+    {
+        return false;
+    }
+
+
+    // Loop over all devices
+    for (;;)
+    {
+        ULONG uReturned = 0;
+        hr = data.pEnumDevices->Next(10000, static_cast<ULONG>(data.pDevices.size()), data.pDevices.data(), &uReturned);
+        if (FAILED(hr))
+        {
+            return false;
+        }
+        if (uReturned == 0)
+            break;
+
+        for (size_t iDevice = 0; iDevice < uReturned; ++iDevice)
+        {
+            // For each device, get its device ID
+            hr = data.pDevices[iDevice]->Get(data.bstrDeviceID, 0L, &data.var, nullptr, nullptr);
+            if (SUCCEEDED(hr) && data.var.vt == VT_BSTR && data.var.bstrVal != nullptr)
+            {
+                // Check if the device ID contains "IG_".  If it does, then it's an XInput device
+                // This information cannot be found from DirectInput
+                if (wcsstr(data.var.bstrVal, L"IG_"))
+                {
+                    // If it does, then get the VID/PID from var.bstrVal
+                    DWORD  dwPid  = 0;
+                    DWORD  dwVid  = 0;
+                    WCHAR* strVid = wcsstr(data.var.bstrVal, L"VID_");
+                    if (strVid && swscanf_s(strVid, L"VID_%4X", &dwVid) != 1)
+                        dwVid = 0;
+                    WCHAR* strPid = wcsstr(data.var.bstrVal, L"PID_");
+                    if (strPid && swscanf_s(strPid, L"PID_%4X", &dwPid) != 1)
+                        dwPid = 0;
```

Comment:
This seems a weird way to retrieve a VID and PID from a device?

Reply:
> This seems a weird way to retrieve a VID and PID from a device?

That's how it's done in the Microsoft Documentation I linked above.

Reply:
looks like msdn suggests this, nvm.